napari/_vispy/overlays/base.py

```diff
@@ -102,7 +125,48 @@ def reset(self):
 
 
 class VispySceneOverlay(VispyBaseOverlay):
-    def __init__(self, *, viewer, **kwargs):
-        super().__init__(**kwargs)
-        self.viewer = viewer
+    """
+    Vispy overlay backend for overlays that live in scene (2D or 3D) space.
+    """
+
+    def __init__(self, *, overlay, node, parent=None):
+        super().__init__(overlay=overlay, node=node, parent=None)
         self.node.transform = MatrixTransform()
+
+
+class LayerOverlayMixin:
+    def __init__(self, *, layer, overlay, node, parent=None):
+        super().__init__(
+            node=node,
+            overlay=overlay,
+            parent=parent,
+        )
+        self.layer = layer
+        self.layer._overlays.events.removing.connect(self._close_if_removed)
+
+    def _close_if_removed(self, event):
+        if self.layer._overlays[event.key] is self:
+            self.close()
+
+    def close(self):
+        disconnect_events(self.layer.events, self)
+        super().close()
+
+
+class ViewerOverlayMixin:
+    def __init__(self, *, viewer, overlay, node, parent=None):
+        super().__init__(
+            node=node,
+            overlay=overlay,
+            parent=parent,
+        )
+        self.viewer = viewer
+        self.viewer._overlays.events.removing.connect(self._close_if_removed)
+
+    def _close_if_removed(self, event):
+        if self.viewer._overlays[event.key] is self:
+            self.close()
```

Comment:
Could you elaborate here? 
The function name suggests that it closes visual if it's removed but we check here if there is a reference stored, and closes only when a reference is stored.

Reply:
So maybe change the name to `_close_if_removing`?